File: src/types/simulation.ts
import type { CreatureGenome, Vector3 } from './genome';
import type { ActivationType } from '../neural/activations';

export enum SimulationPhase {
  MENU = 'menu',
  GENERATING = 'generating',
  SIMULATING = 'simulating',
  DISPLAYING = 'displaying',
  EVOLVING = 'evolving',
  PAUSED = 'paused'
}

export interface SimulationConfig {
  // Physics
  gravity: number;              // -9.8 to -30
  ground_friction: number;       // 0.3 to 1.0
  time_step: number;             // Physics timestep (1/physics_fps)
  physics_fps: number;           // Physics frames per second (15-120)
  simulation_duration: number;   // Seconds per generation

  // Muscle constraints
  muscle_velocity_cap: number;       // Max muscle length change per second (0.1-20.0)
  muscle_damping_multiplier: number; // Global damping scale (0.1-5.0, higher = more resistance)
  max_extension_ratio: number;       // Max muscle stretch (1.2-5.0, 2.0 = 50%-200% of rest length)

  // Evolution
  population_size: number;       // 100
  cull_percentage: number;       // 0.5 (bottom 50%)
  selection_method: 'truncation' | 'tournament' | 'rank' | 'speciation';  // How survivors are selected
  tournament_size: number;       // For tournament selection: number of contestants per round
  mutation_rate: number;         // Per-gene mutation rate (0.1-0.5)
  mutation_magnitude: number;    // How much values change
  crossover_rate: number;        // Probability of crossover vs mutation for new creatures (0-1)
  elite_count: number;           // Deprecated - survivors determined by cull_percentage
  use_crossover: boolean;        // Whether to use crossover (vs clone-only)

  // Creature constraints
  min_nodes: number;
  max_nodes: number;
  max_muscles: number;
  max_allowed_frequency: number;  // Max muscle frequency before creature is penalized

  // Environment
  pellet_count: number;          // Number of pellets per arena
  arena_size: number;            // Size of simulation arena

  // Fitness function (simple model)
  fitness_pellet_points: number;          // Points per pellet collected (default 20, on top of 80 progress = 100 total)
  fitness_progress_max: number;           // Max points for progress toward pellet (default 80)
  fitness_distance_per_unit: number;       // Points per unit of distance traveled (default 3)
  fitness_distance_traveled_max: number;   // Max points for distance traveled (default 20)
  fitness_regression_penalty: number;     // Max penalty for moving away after 1st pellet collection (default 20)

  // Neural network settings (neuroevolution)
  use_neural_net: boolean;              // Enable neural network control
  neural_mode: 'hybrid' | 'pure' | 'neat';  // How NN output is used (neat = variable topology)
  bias_mode: 'none' | 'node' | 'bias_node';  // Bias implementation: none, per-node attribute, or bias input node
  time_encoding: 'none' | 'cyclic' | 'sin' | 'raw' | 'sin_raw';  // Time encoding (default: none for pure, cyclic for hybrid)
  neural_hidden_size: number;           // Neurons in hidden layer
  neural_activation: ActivationType;   // Activation function
  weight_mutation_rate: number;         // Target/end probability each weight mutates
  weight_mutation_magnitude: number;    // Std dev of weight perturbation
  weight_mutation_decay: 'off' | 'linear' | 'exponential';  // Mutation rate decay mode
  neural_output_bias: number;           // Initial output neuron bias (0 = neutral, inside dead zone)
  fitness_efficiency_penalty: number;   // Penalty per unit of total muscle activation (encourages efficient movement)
  neural_dead_zone: number;             // Dead zone threshold for pure mode (outputs < threshold become 0)
  neural_update_hz: number;             // NN update frequency in Hz (5-60, default 15)
  output_smoothing_alpha: number;       // Exponential smoothing factor (0.05-1.0, 1.0 = no smoothing)

  // Adaptive mutation (escape stagnation)
  use_adaptive_mutation: boolean;       // Auto-boost mutation when fitness stagnates
  stagnation_threshold: number;        // Window size for comparing fitness (default 20)
  adaptive_mutation_boost: number;      // Multiplier for mutation rate during stagnation (e.g., 2.0 = double)
  max_adaptive_boost: number;           // Maximum boost multiplier cap (e.g., 8.0)
  improvement_threshold: number;       // Minimum fitness improvement to count as progress (absolute points)

  // Crossover method for neural weights
  neural_crossover_method: 'interpolation' | 'uniform' | 'sbx';  // How neural weights are combined during crossover
  sbx_eta: number;                     // SBX distribution index (0.5-5.0, lower = more exploration)

  // Fitness sharing (diversity maintenance)
  use_fitness_sharing: boolean;         // Enable fitness sharing to penalize similar creatures
  sharing_radius: number;              // Genome distance threshold for sharing (0.1-2.0)

  // Speciation parameters (used when selection_method='speciation')
  compatibility_threshold: number;     // Genome distance threshold for same species (0.1-3.0)
  min_species_size: number;             // Minimum survivors per species (1-20)

  // NEAT (NeuroEvolution of Augmenting Topologies) - configured when neural_mode === 'neat'
  neat_initial_connectivity: 'full' | 'sparse_inputs' | 'sparse_outputs' | 'none';  // Initial network connectivity
  neat_add_connection_rate: number;      // Probability to add a new connection (0.01-0.2)
  neat_add_node_rate: number;            // Probability to add a new hidden node (0.01-0.1)
  neat_enable_rate: number;             // Probability to re-enable a disabled connection (0.01-0.1)
  neat_disable_rate: number;            // Probability to disable an enabled connection (0.01-0.1)
  neat_excess_coefficient: number;      // Weight for excess genes in compatibility distance (0-10)
  neat_disjoint_coefficient: number;    // Weight for disjoint genes in compatibility distance (0-10)
  neat_weight_coefficient: number;      // Weight for weight differences in compatibility distance (0-10)
  neat_max_hidden_nodes: number;         // Maximum hidden neurons to prevent bloat (1-128)

  // Proprioception (body-sensing inputs)
  use_proprioception: boolean;         // Enable body-sensing inputs (muscle strain, node velocities, ground contact)
  proprioception_inputs: 'strain' | 'velocity' | 'ground' | 'all';  // Which proprioception inputs to include

  // Frame storage mode for replay capability
  frame_storage_mode: 'none' | 'sparse' | 'all';  // none = no replays, sparse = top/bottom only, all = all creatures
  sparse_top_count: number;             // Number of top performers to store frames for (sparse mode)
  sparse_bottom_count: number;          // Number of bottom performers to store frames for (sparse mode)
}

export const DEFAULT_CONFIG: SimulationConfig = {
  gravity: -9.8,
  ground_friction: 0.5,
  time_step: 1 / 30,
  physics_fps: 30,
  simulation_duration: 20,

  // Muscle constraints
  muscle_velocity_cap: 5.0,       // Max muscle length change per second
  muscle_damping_multiplier: 1.0, // Default: no scaling (higher = more 'underwater' feel)
  max_extension_ratio: 2.0,       // Muscles can stretch 50%-200% of rest length

  population_size: 100,
  cull_percentage: 0.5,
  selection_method: 'rank',
  tournament_size: 3,
  mutation_rate: 0.2,
  mutation_magnitude: 0.3,
  crossover_rate: 0.5,
  elite_count: 5,
  use_crossover: true,

  min_nodes: 3,
  max_nodes: 8,
  max_muscles: 15,
  max_allowed_frequency: 3.0,

  pellet_count: 3,
  arena_size: 10,

  // Fitness function defaults
  fitness_pellet_points: 20,    // On top of 80 progress = 100 total per pellet
  fitness_progress_max: 80,
  fitness_distance_per_unit: 3,
  fitness_distance_traveled_max: 20,
  fitness_regression_penalty: 20,

  // Neural network defaults (enabled by default)
  use_neural_net: true,
  neural_mode: 'pure',
  bias_mode: 'node',      // Default: per-node biases (NEAT will auto-switch to 'bias_node')
  time_encoding: 'none',  // Default 'none' for pure mode; 'cyclic' recommended for hybrid
  neural_hidden_size: 8,
  neural_activation: 'tanh',
  weight_mutation_rate: 0.2,
  weight_mutation_magnitude: 0.05,
  weight_mutation_decay: 'linear',
  neural_output_bias: -0.1,        // Slight negative bias so muscles must evolve to activate
  fitness_efficiency_penalty: 0.1, // Subtle penalty for excessive activation
  neural_dead_zone: 0.1,           // Outputs with absolute value < 0.1 become 0 in pure mode
  neural_update_hz: 10,            // Update NN 10 times per second (smoother movement)
  output_smoothing_alpha: 0.15,    // Aggressive smoothing (0.15 = 15% new + 85% old)

  // Adaptive mutation defaults
  use_adaptive_mutation: false,    // Off by default - enable for long runs
  stagnation_threshold: 20,       // 20-generation window for comparison
  adaptive_mutation_boost: 2.0,    // Double mutation rate during stagnation
  max_adaptive_boost: 8.0,         // Cap at 8x mutation rate
  improvement_threshold: 5.0,     // Need 5+ fitness points improvement to count

  // Crossover method defaults
  neural_crossover_method: 'sbx',  // SBX produces smoother offspring distribution
  sbx_eta: 2.0,                   // Balanced exploration/exploitation

  // Fitness sharing defaults
  use_fitness_sharing: false,      // Off by default - enable to maintain diversity
  sharing_radius: 0.5,            // Moderate sharing radius

  // Speciation parameters (used when selection_method='speciation')
  compatibility_threshold: 1.0,   // Genome distance threshold for same species
  min_species_size: 2,             // Minimum survivors per species

  // NEAT defaults (used when neural_mode === 'neat')
  neat_initial_connectivity: 'full',  // Start with all inputs connected to all outputs
  neat_add_connection_rate: 0.05,   // 5% chance to add connection per genome
  neat_add_node_rate: 0.03,         // 3% chance to add node per genome
  neat_enable_rate: 0.02,          // 2% chance to re-enable disabled connection
  neat_disable_rate: 0.01,         // 1% chance to disable enabled connection
  neat_excess_coefficient: 1.0,    // Standard weight for excess genes
  neat_disjoint_coefficient: 1.0,  // Standard weight for disjoint genes
  neat_weight_coefficient: 0.4,    // Lower weight for weight differences (per NEAT paper)
  neat_max_hidden_nodes: 16,        // Reasonable limit for complexity

  // Proprioception defaults
  use_proprioception: false,      // Off by default - experimental feature
  proprioception_inputs: 'all',   // Include all proprioception inputs when enabled

  // Frame storage mode
  frame_storage_mode: 'all',       // Store frames for all creatures by default (enables replays)
  sparse_top_count: 10,            // Store frames for top 10 performers in sparse mode
  sparse_bottom_count: 5,          // Store frames for bottom 5 performers in sparse mode
};

export interface CreatureState {
  genome: CreatureGenome;

  // Physics state
  nodePositions: Map<string, Vector3>;
  nodeVelocities: Map<string, Vector3>;
  centerOfMass: Vector3;

  // Simulation results
  fitness: number;
  pelletsCollected: number;
  distanceTraveled: number;
  closestPelletDistance: number;

  // Grid position for display
  gridX: number;
  gridY: number;

  // Selection state
  isSelected: boolean;
}

export interface PelletState {
  id: string;
  position: Vector3;
  collected: boolean;
}

export interface FitnessHistoryEntry {
  generation: number;
  best: number;
  average: number;
  worst: number;
}

export interface ConfigHistoryEntry {
  generation: number;
  avgNodes: number;
  avgMuscles: number;
  avgBodySize: number;
}

export interface PopulationStats {
  generation: number;
  bestFitness: number;
  averageFitness: number;
  worstFitness: number;
  avgNodes: number;
  avgMuscles: number;
}

// =============================================================================
// Simulation Result Types (shared by frontend and backend)
// =============================================================================

export interface SimulationFrame {
  time: number;
  nodePositions: Map<string, Vector3>;
  centerOfMass: Vector3;
  activePelletIndex: number;
}

export type DisqualificationReason =
  | null  // Not disqualified
  | 'frequency_exceeded'  // Muscle frequency too high
  | 'physics_explosion'   // Creature flew off into space
  | 'nan_position';       // Position became NaN

export interface CreatureSimulationResult {
  genome: CreatureGenome;
  frames: SimulationFrame[];
  finalFitness: number;
  pelletsCollected: number;
  distanceTraveled: number;
  netDisplacement: number;  // Straight-line distance from start to end (metric only, not fitness)
  closestPelletDistance: number;
  pellets: PelletData[];
  fitnessOverTime: number[];
  disqualified: DisqualificationReason;
  activationsPerFrame?: FrameActivations[];  // Full neural network activations per frame

  // UI-specific properties (set by frontend from evolution step response)
  _isSurvivor?: boolean;   // For animation: creature survived from previous generation
  _hasFrames?: boolean;    // For replay button: frames are available in database

  // Lifecycle tracking (from API)
  birthGeneration?: number;  // Generation this creature was created
  deathGeneration?: number;  // Generation this creature died (null = still alive)
}

export interface PelletData {
  id: string;
  position: Vector3;
  collectedAtFrame: number | null;
  spawnedAtFrame: number;
  initialDistance: number;  // Distance from creature edge when pellet spawned
}

/**
 * Full neural network activation data for a single frame.
 * Includes inputs, hidden layer activations, and outputs.
 */
export interface FrameActivations {
  inputs: number[];   // Sensor inputs fed to network
  hidden: number[];   // Hidden layer activations
  outputs: number[];  // Muscle outputs (post-dead-zone)
  outputs_raw?: number[];  // Raw muscle outputs before dead zone (pure mode only)
}

// =============================================================================
// NEAT (NeuroEvolution of Augmenting Topologies) Types
// =============================================================================

/**
 * NEAT neuron gene - represents a node in the neural network.
 */
export interface NeuronGene {
  id: number;                                    // Unique neuron ID
  type: 'input' | 'hidden' | 'output' | 'bias';  // Neuron type (bias = always 1.0)
  bias: number;                                  // Bias value (0 for input/bias neurons)
  innovation?: number;                           // Innovation number (only for hidden neurons)
}

/**
 * NEAT connection gene - represents a connection between neurons.
 */
export interface ConnectionGene {
  fromNode: number;      // Source neuron ID
  toNode: number;        // Target neuron ID
  weight: number;        // Connection weight
  enabled: boolean;      // Whether connection is active
  innovation: number;    // Innovation number for gene alignment
}

/**
 * NEAT genome - variable topology neural network.
 */
export interface NEATGenome {
  neurons: NeuronGene[];        // All neurons (input, hidden, output)
  connections: ConnectionGene[];  // All connections
  activation: string;           // Activation function ('tanh', 'relu', etc.)
}

/**
 * Innovation counter state - persisted across generations for NEAT runs.
 */
export interface InnovationCounterState {
  nextConnection: number;  // Next connection innovation number
  nextNode: number;        // Next node innovation number
}

// =============================================================================
// Migration Utilities
// =============================================================================

/**
 * Migrate old configs that used useNEAT boolean to neural_mode='neat'.
 * Call this when loading configs from storage or API.
 */
export function migrateConfig(config: Partial<SimulationConfig> & { useNEAT?: boolean }): Partial<SimulationConfig> {
  const migrated = { ...config };
  if ('useNEAT' in migrated && migrated.useNEAT) {
    migrated.neural_mode = 'neat';
  }
  delete (migrated as { useNEAT?: boolean }).useNEAT;
  return migrated;
}

File: backend/app/schemas/simulation.py
"""
API schemas for simulation endpoints.

Matches the internal PyTorch simulation config (app/simulation/config.py)
while providing Pydantic validation for API requests.
"""

from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

from app.schemas.genome import CreatureGenome


class SimulationConfig(BaseModel):
    """
    Configuration for running a simulation.

    Matches TypeScript SimulationConfig interface.
    All fields have sensible defaults so clients can omit most of them.
    """

    # Physics
    gravity: float = Field(default=-9.8, ge=-30.0, le=-1.0)
    ground_friction: float = Field(default=0.5, ge=0.3, le=1.0)
    time_step: float = Field(default=1/30, ge=1/120, le=1/15)  # 15-120 FPS, default 30 FPS for speed
    simulation_duration: float = Field(default=20.0, ge=1.0, le=60.0)

    # Muscle constraints
    muscle_velocity_cap: float = Field(default=5.0, ge=0.1, le=20.0)  # Max muscle length change per second
    muscle_damping_multiplier: float = Field(default=1.0, ge=0.1, le=5.0)  # Global damping scale (higher = more resistance)
    max_extension_ratio: float = Field(default=2.0, ge=1.2, le=5.0)  # Max muscle stretch (2.0 = 50%-200% of rest length)

    # Evolution (not used in simulation, but included for parity)
    population_size: int = Field(default=100, ge=10, le=10000)
    cull_percentage: float = Field(default=0.5, ge=0.1, le=0.9)
    selection_method: Literal['truncation', 'tournament', 'rank', 'speciation'] = 'rank'
    tournament_size: int = Field(default=3, ge=2, le=10)
    mutation_rate: float = Field(default=0.2, ge=0.0, le=1.0)
    mutation_magnitude: float = Field(default=0.3, ge=0.0, le=1.0)
    crossover_rate: float = Field(default=0.5, ge=0.0, le=1.0)
    elite_count: int = Field(default=5, ge=0, le=100)
    use_crossover: bool = True

    # Creature constraints
    min_nodes: int = Field(default=3, ge=2, le=20)
    max_nodes: int = Field(default=8, ge=2, le=20)
    max_muscles: int = Field(default=15, ge=1, le=50)
    max_allowed_frequency: float = Field(default=3.0, ge=0.5, le=10.0)

    # Environment
    pellet_count: int = Field(default=3, ge=0, le=50)
    arena_size: float = Field(default=10.0, ge=5.0, le=100.0)

    # Fitness function weights
    fitness_pellet_points: float = Field(default=20.0, ge=0.0)  # On top of 80 progress = 100 total
    fitness_progress_max: float = Field(default=80.0, ge=0.0)
    fitness_distance_per_unit: float = Field(default=3.0, ge=0.0)
    fitness_distance_traveled_max: float = Field(default=20.0, ge=0.0)
    fitness_regression_penalty: float = Field(default=20.0, ge=0.0)  # Only after first collection

    # Neural network settings
    use_neural_net: bool = True
    neural_mode: Literal['hybrid', 'pure', 'neat'] = 'pure'
    bias_mode: Literal['none', 'node', 'bias_node'] = 'node'  # Bias implementation: none, per-node attribute, or bias input node
    time_encoding: Literal['none', 'cyclic', 'sin', 'raw', 'sin_raw'] = 'none'  # Time encoding (default: none for pure, cyclic recommended for hybrid)
    neural_hidden_size: int = Field(default=8, ge=1, le=64)
    neural_activation: str = Field(default='tanh')
    weight_mutation_rate: float = Field(default=0.2, ge=0.0, le=1.0)
    weight_mutation_magnitude: float = Field(default=0.05, ge=0.0, le=2.0)
    weight_mutation_decay: Literal['off', 'linear', 'exponential'] = 'linear'
    neural_output_bias: float = Field(default=-0.1, ge=-2.0, le=2.0)
    fitness_efficiency_penalty: float = Field(default=0.1, ge=0.0, le=5.0)
    neural_dead_zone: float = Field(default=0.1, ge=0.0, le=1.0)
    neural_update_hz: int = Field(default=10, ge=5, le=60)  # NN update frequency (Hz)
    output_smoothing_alpha: float = Field(default=0.15, ge=0.05, le=1.0)  # Exponential smoothing (1.0 = no smoothing)

    # Adaptive mutation
    use_adaptive_mutation: bool = False
    stagnation_threshold: int = Field(default=20, ge=5, le=100)
    adaptive_mutation_boost: float = Field(default=2.0, ge=1.1, le=5.0)
    max_adaptive_boost: float = Field(default=8.0, ge=2.0, le=32.0)
    improvement_threshold: float = Field(default=5.0, ge=1.0, le=50.0)

    # Crossover method for neural weights
    neural_crossover_method: Literal['interpolation', 'uniform', 'sbx'] = 'sbx'
    sbx_eta: float = Field(default=2.0, ge=0.5, le=5.0)

    # Fitness sharing (diversity maintenance)
    use_fitness_sharing: bool = False
    sharing_radius: float = Field(default=0.5, ge=0.1, le=2.0)

    # Speciation parameters (used when selection_method='speciation')
    compatibility_threshold: float = Field(default=1.0, ge=0.1, le=3.0)
    min_species_size: int = Field(default=2, ge=1, le=20)

    # NEAT (NeuroEvolution of Augmenting Topologies) - configured when neural_mode == 'neat'
    neat_initial_connectivity: Literal['full', 'sparse_inputs', 'sparse_outputs', 'none'] = 'full'  # Initial network connectivity
    neat_add_connection_rate: float = Field(default=0.05, ge=0.0, le=1.0)  # Probability to add connection
    neat_add_node_rate: float = Field(default=0.03, ge=0.0, le=1.0)  # Probability to add node
    neat_enable_rate: float = Field(default=0.02, ge=0.0, le=1.0)  # Probability to re-enable connection
    neat_disable_rate: float = Field(default=0.01, ge=0.0, le=1.0)  # Probability to disable connection
    neat_excess_coefficient: float = Field(default=1.0, ge=0.0, le=10.0)  # Weight for excess genes in distance
    neat_disjoint_coefficient: float = Field(default=1.0, ge=0.0, le=10.0)  # Weight for disjoint genes in distance
    neat_weight_coefficient: float = Field(default=0.4, ge=0.0, le=10.0)  # Weight for weight differences in distance
    neat_max_hidden_nodes: int = Field(default=16, ge=1, le=128)  # Maximum hidden neurons to prevent bloat

    # Proprioception (body-sensing inputs)
    use_proprioception: bool = False
    proprioception_inputs: Literal['strain', 'velocity', 'ground', 'all'] = 'all'

    # Disqualification thresholds
    position_threshold: float = Field(default=50.0, ge=10.0)
    height_threshold: float = Field(default=30.0, ge=5.0)

    # Internal settings
    pellet_collection_radius: float = Field(default=0.75, ge=0.1, le=5.0)
    max_pellet_distance: float = Field(default=20.0, ge=5.0, le=100.0)

    # Frame recording
    # frame_storage_mode: 'none' (default), 'all' (UI/--store), 'sparse' (--sparse-store: top 10 + bottom 10)
    frame_storage_mode: Literal['none', 'all', 'sparse'] = 'none'
    frame_rate: int = Field(default=15, ge=1, le=60)
    sparse_top_count: int = Field(default=10, ge=1, le=50)
    sparse_bottom_count: int = Field(default=10, ge=1, le=50)

    @model_validator(mode='before')
    @classmethod
    def migrate_and_enforce_neat_defaults(cls, data: Any) -> Any:
        """Migrate legacy fields and enforce NEAT defaults."""
        if isinstance(data, dict):
            # Check for legacy use_neat or useNEAT fields
            use_neat = data.pop('use_neat', None) or data.pop('useNEAT', None)
            if use_neat:
                data['neural_mode'] = 'neat'

            # Migrate legacy use_speciation to selection_method
            use_speciation = data.pop('use_speciation', None) or data.pop('useSpeciation', None)
            if use_speciation:
                data['selection_method'] = 'speciation'

            # ENFORCE NEAT defaults - these are REQUIRED for NEAT to work properly
            # Speciation protects structural innovations until weights adapt
            if data.get('neural_mode') == 'neat':
                data['selection_method'] = 'speciation'  # REQUIRED - without this, topology won't evolve
                data['use_fitness_sharing'] = False  # Redundant with speciation
                # Default to bias_node for NEAT (more faithful to original NEAT)
                if 'bias_mode' not in data:
                    data['bias_mode'] = 'bias_node'
        return data

    @property
    def record_frames(self) -> bool:
        """Backwards-compatible property: True if any frame storage is enabled."""
        return self.frame_storage_mode != 'none'

    class Config:
        extra = 'ignore'  # Ignore unknown fields for forward compatibility


class FitnessBreakdown(BaseModel):
    """Detailed breakdown of fitness components."""

    pellet_points: float = 0.0
    progress: float = 0.0
    distance_traveled: float = 0.0
    regression_penalty: float = 0.0
    efficiency_penalty: float = 0.0


class PelletResult(BaseModel):
    """Data about a pellet's position and collection state."""

    id: str
    position: dict  # {x, y, z}
    collected_at_frame: int | None = None
    spawned_at_frame: int
    initial_distance: float


class SimulationResult(BaseModel):
    """Result from simulating a single creature."""

    genome_id: str
    fitness: float
    pellets_collected: int
    disqualified: bool
    disqualified_reason: str | None = None

    # Fitness breakdown (optional, for detailed analysis)
    fitness_breakdown: FitnessBreakdown | None = None

    # Movement metrics
    net_displacement: float = 0.0
    distance_traveled: float = 0.0
    total_activation: float = 0.0  # For neural network efficiency

    # Frame data (only if record_frames=True)
    frame_count: int = 0
    frames: list | None = None  # Flattened frame data if requested

    # Pellet data (for replay visualization)
    pellets: list[PelletResult] | None = None

    # Fitness over time (for graphing actual fitness progression)
    fitness_over_time: list[float] | None = None

    # Neural network activations per frame (full forward pass data)
    # Only populated for neural controller creatures when frames are recorded
    # Each entry: { "inputs": [...], "hidden": [...], "outputs": [...] }
    activations_per_frame: list[dict] | None = None


class BatchSimulationRequest(BaseModel):
    """Request to simulate a batch of creatures."""

    genomes: list[CreatureGenome]
    config: SimulationConfig = Field(default_factory=SimulationConfig)


class BatchSimulationResponse(BaseModel):
    """Response from batch simulation."""

    results: list[SimulationResult]
    total_time_ms: int
    creatures_per_second: float
